OSDOCS-3147: Adding EUS to EUS upgrade doc (#40619)
openshift/openshift-docs: modules/updating-eus-to-eus-upgrade.adoc

// Module included in the following assemblies:
//
// * updating/preparing-eus-eus-upgrade.adoc

[id="updating-eus-to-eus-upgrade_{context}"]
= EUS to EUS upgrade

The following procedure pauses all non-master MachineConfigPools and performs upgrades from 4.8 to 4.9 to 4.10, then unpauses the previously paused MachineConfigPools.
Following this procedure reduces the total upgrade duration and the number of times worker nodes are restarted.

.Prerequisites

* Review the release notes for {product-title} 4.9 and 4.10

> * Review the release notes and product lifecycles for any layered products and OLM Operators, some may require updates either before or during an EUS to EUS upgrade.

Reviewer: Should be two sentences (period after "OLM Operators." Or a semi-colon, if you like those things. :-)

* Ensure that you are familiar with version-specific prerequisites, such as link:https://docs.openshift.com/container-platform/4.9/updating/updating-cluster-prepare.html#update-preparing-migrate_updating-cluster-prepare[administrator acknowledgement] that is required prior to upgrading from {product-title} 4.8 to 4.9.
* Verify that your cluster is not running {product-title} 4.8.14 version or earlier.
If your cluster is running {product-title} 4.8.14 or earlier, you must upgrade to a later 4.8.z version prior to upgrading to 4.9.

> The upgrade to 4.8.14 or later which is necessary to fulfill the minimum version requirements that must be performed without pausing MachineConfigPools.

Reviewer 2: ```suggestion
The upgrade to 4.8.14 or later which is necessary to fulfill the minimum version requirements must be performed without pausing MachineConfigPools.
```

> The upgrade to 4.8.14 or later which is necessary to fulfill the minimum version requirements that must be performed without pausing MachineConfigPools.

Reviewer: * Line 17: Is "z" necessary in "4.8.z?" It's already specified that a version later than 4.8.14 is needed.
* Line 18: Is a negative overlap created when saying to upgrade to 4.8.14 or later when Line 17 says that 4.8.14 won't work?

Author: I adjusted the wording across the board to help with this issue.

* Verify that MachineConfigPools is unpaused.

.Procedure

. Upgrade any OLM Operators to versions that are compatible with both versions you are upgrading to.

. Verify that all MachineConfigPools display a status of `UPDATED` and no MachineConfigPools display a status of `UPDATING`.
To view the status of all MachineConfigPools, run the following command:
+
[source,terminal]
----
$ oc get mcp
----
+
.Example output
+
Output is trimmed for clarity:
+
[source,terminal]
----
NAME     CONFIG                                         	UPDATED   UPDATING
master   rendered-master-ecbb9582781c1091e1c9f19d50cf836c       True  	  False
worker   rendered-worker-00a3f0c68ae94e747193156b491553d5       True  	  False
----

. To pause the MachineConfigPools you wish to skip reboots on, run the following commands:
+
[NOTE]
====
You cannot pause the master pool.
====
+
[source,terminal]
----
$ oc patch mcp/worker --type merge --patch '{"spec":{"paused":true}}'
----

. To change to the `eus-4.10` channel and upgrade to 4.9, run the following commands:
+
[source,terminal]
----
$ oc adm upgrade channel eus-4.10
$ oc adm upgrade --to-latest
----
+
.Example output
+
[source,terminal]
----

> Updating to latest version 4.9.11

Reviewer 2: ```suggestion
Updating to latest version 4.9.18
```

Reviewer 2: We've recently become aware of an issue which required us to bump the minimum version required to upgrade from 4.9 to 4.10. This probably doesn't matter but I don't want someone seeing 4.9.11 here and questioning why they can't upgrade to 4.10 from 4.9.11.

----

> . To ensure the 4.9 upgrades is completed successfully retrieve the cluster version, run the following command:

Reviewer: Ooops. Missed this the first runthrough. "Is" should be "are."

+
[source,terminal]
----
$ oc get clusterversion
----
+
.Example output
+
[source,terminal]
----
NAME  	  VERSION  AVAILABLE  PROGRESSING   SINCE   STATUS

> version   4.9.11   True       False         6m29s   Cluster version is 4.9.11

Reviewer 2: ```suggestion
version   4.9.18   True       False         6m29s   Cluster version is 4.9.18
```

----

. If necessary, upgrade OLM operators using the Administrator perspective on the web console.

. To upgrade to 4.10, run the following command:
+
[source,terminal]
----
$ oc adm upgrade --to-latest
----

. To ensure the 4.10 upgrade is completed successfully retrieve the cluster version, run the following command:
+
[source,terminal]
----
$ oc get clusterversion
----
+
.Example output
+
[source,terminal]
----
NAME  	  VERSION  AVAILABLE  PROGRESSING   SINCE   STATUS
version   4.10.1   True       False         6m29s   Cluster version is 4.10.1
----

. To unpause all previously paused MachineConfigPools, run the following command:
+
[source,terminal]
----
$ oc patch mcp/worker --type merge --patch '{"spec":{"paused":false}}'
----
+
[NOTE]
====
If pools are not unpaused, the cluster is not permitted to upgrade to any future minors and maintenance tasks such as certificate rotation are inhibited. This puts the cluster at risk for future degradation.
====

> . To verify that your previously paused pools have updated and you cluster completed the upgrade to 4.10, run the following command:

Reviewer: Should "you cluster" be "your cluster?"

// Module included in the following assemblies:
//
// * updating/preparing-eus-eus-upgrade.adoc

[id="updating-eus-to-eus-upgrade_{context}"]
= EUS to EUS upgrade

The following procedure pauses all non-master MachineConfigPools and performs upgrades from 4.8 to 4.9 to 4.10, then unpauses the previously paused MachineConfigPools.

> Following this procedure reduces the total upgrade duration and the number of times workloads are restarted.

Reviewer 3: s/workloads/worker nodes/

.Prerequisites

> * Review the release notes for {product-title} 4.9 and 4.10

Reviewer 2: ```suggestion
* Review the release notes for {product-title} 4.9 and 4.10
* Review the release notes and product lifecycles for any layered products and OLM Operators, some may require updates either before or during an EUS-to-EUS upgrade
```

> * Ensure that you are familiar with version-specific prerequisites, such as Administrator Acknowledgement that is required prior to upgrading from {product-title} 4.8 to 4.9.

Reviewer 3: Should we link the admin doc URL here?

Author: Add link: https://docs.openshift.com/container-platform/4.9/updating/updating-cluster-prepare.html#update-preparing-migrate_updating-cluster-prepare

* Verify that your cluster is not running {product-title} 4.8.14 version or earlier.
If your cluster is running {product-title} 4.8.14 or earlier, you must upgrade to a later 4.8.z version prior to upgrading to 4.9.

> The upgrade to 4.8.14 or later fulfills the minimum version requirements that must be performed without pausing MachineConfigPools.

Reviewer 2: ```suggestion
The upgrade to 4.8.14 or later which is necessary to fulfill the minimum version requirements must be performed without pausing MachineConfigPools.
```
Or something like that, just a bit hard to follow as is.

* Verify that MachineConfigPools is unpaused.

.Procedure

> 

Reviewer 2: ```suggestion
. Upgrade any OLM Operators to versions which are compatible with both versions you will be upgrading to.
```

. Verify that all MachineConfigPools display a status of `UPDATED` and no MachineConfigPools display a status of `UPDATING`.
To view the status of all MachineConfigPools, run the following command:
+
[source,terminal]
----
$ oc get mcp
----
+

> .Example output

Reviewer 4: This line seems indented. Refer to https://github.com/openshift/openshift-docs/blob/840ecfcd8214b9109d0abfd415e488c0fdcd53c5/modules/updating-eus-to-eus-upgrade.adoc

Author: @jiajliu Yes, it is. It is meant to group things together.

Reviewer 4: ack

+
Output is trimmed for clarity:
+
[source,terminal]
----
NAME     CONFIG                                         	UPDATED   UPDATING
master   rendered-master-ecbb9582781c1091e1c9f19d50cf836c       True  	  False
worker   rendered-worker-00a3f0c68ae94e747193156b491553d5       True  	  False
----

. To pause the MachineConfigPools you wish to skip reboots on, run the following commands:
+
[NOTE]
====
You cannot pause the master pool.
====
+
[source,terminal]
----

> $ oc patch mcp/worker --type merge --patch '{"spec":{"paused":true}}'

Reviewer 4: How about add the output of the patch command?
```
# ./oc patch mcp/worker --type merge --patch '{"spec":{"paused":true}}'
machineconfigpool.machineconfiguration.openshift.io/worker patched
```

> $ oc get mcp -o

Reviewer 4: The command is split to two lines wrongly.

Author: @jiajliu not sure what you mean? @sdodson is this not how it Is supposed to be?

Reviewer 4: Yes, i mean the command is split to line 54 and 64, which should be together in one line.

----
+
.Example output
+
[source,terminal]
----
custom-columns=NAME:.metadata.name,PAUSED:.spec.paused
NAME 	    PAUSED
master      false
worker      true
----

. To change to the `eus-4.10` channel and upgrade to 4.9, run the following commands:
+
[source,terminal]
----
$ oc adm upgrade channel eus-4.10

> $ oc adm upgrade --to-latest

Reviewer 4: When it's a disconnected cluster, we can not do upgrade with --to-latest directly. Maybe we could give a link here for users to redirect to upgrade in disconnected cluster section.

Reviewer 4: make sense. ack.

----
+
.Example output
+
[source,terminal]
----
Updating to latest version 4.9.11
----

. To ensure the 4.9 upgrades is completed successfully retrieve the cluster version, run the following command:
+
[source,terminal]
----
$ oc get clusterversion
----
+
.Example output
+
[source,terminal]
----
NAME  	  VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.9.11	  True    	  False     	  6m29s   Cluster version is 4.9.11
----

. To upgrade to 4.10, run the following command:
+
[source,terminal]
----
$ oc adm upgrade --to-latest
----

. To ensure the 4.10 upgrade is completed successfully retrieve the cluster version, run the following command:
+
[source,terminal]
----
$ oc get clusterversion
----
+
.Example output
+
[source,terminal]
----
NAME  	  VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS

> version   4.10.1	  True    	  False     	  6m29s   Cluster version is 4.10.1

Author: @sdodson - Could you verify this example output? There wasn't any example output in the Google Doc, but I based it off the 4.9.11 above.

Reviewer 2: That's fine.

----

. To unpause all previously paused MachineConfigPools, run the following command:
+
[source,terminal]
----

> $ oc patch mcp/worker --type merge --patch '{"spec":{"paused":false}}'

Reviewer 4: The same here.
```
# ./oc patch mcp/worker --type merge --patch '{"spec":{"paused":false}}'
machineconfigpool.machineconfiguration.openshift.io/worker patched
```

----
+
[NOTE]
====
If pools are not unpaused, the cluster is not permitted to upgrade to any future minors and maintenance tasks such as certificate rotation are inhibited. This puts the cluster at risk for future degradation.
====

. To verify that your previously paused pools have updated and you cluster completed the upgrade to 4.10, run the following command:
+
[source,terminal]
----
$ oc get mcp
----
+
.Example output
+
Output is trimmed for clarity:
+
[source,terminal]
----
NAME 	   CONFIG                                            UPDATED     UPDATING
master   rendered-master-ecbb9582781c1091e1c9f19d50cf836c    True  	 False
worker   rendered-worker-00a3f0c68ae94e747193156b491553d5    True 	 False

> ----

Reviewer 4: We could give a new mc config here(in contrast the old one before update in step 1).  
```
master   rendered-master-52da4d2760807cb2b96a3402179a9a4c
worker   rendered-worker-4756f60eccae96fb9dcb4c392c69d497
```

+
[source,terminal]
----
Updating to latest version 4.9.11
----

. To ensure the 4.9 upgrades is completed successfully retrieve the cluster version, run the following command:
+
[source,terminal]
----
$ oc get clusterversion
----
+
.Example output
+
[source,terminal]
----
NAME  	  VERSION  AVAILABLE  PROGRESSING   SINCE   STATUS
version   4.9.11   True       False         6m29s   Cluster version is 4.9.11
----

. If necessary, upgrade OLM operators using the Administrator perspective on the web console.

. To upgrade to 4.10, run the following command:
+
[source,terminal]
----
$ oc adm upgrade --to-latest
----

. To ensure the 4.10 upgrade is completed successfully retrieve the cluster version, run the following command:
+
[source,terminal]
----
$ oc get clusterversion
----
+
.Example output
+
[source,terminal]
----
NAME  	  VERSION  AVAILABLE  PROGRESSING   SINCE   STATUS
version   4.10.1   True       False         6m29s   Cluster version is 4.10.1
----

. To unpause all previously paused MachineConfigPools, run the following command:
+
[source,terminal]
----
$ oc patch mcp/worker --type merge --patch '{"spec":{"paused":false}}'
----
+
.Example output
+
[source,terminal]
----

> # ./oc patch mcp/worker --type merge --patch '{"spec":{"paused":false}}'

Reviewer 4: @sagidlow Looks like we could remove line 135 if we keep line 128. They are duplicated.
